File: main/client/src/mill/main/client/ProxyStream.java
package mill.main.client;


import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * Logic to capture a pair of streams (typically stdout and stderr), combining
 * them into a single stream, and splitting it back into two streams later while
 * preserving ordering. This is useful for capturing stderr and stdout and forwarding
 * them to a terminal while strictly preserving the ordering, i.e. users won't see
 * exception stack traces and printlns arriving jumbled up and impossible to debug
 *
 * This works by converting writes from either of the two streams into packets of
 * the form:
 *
 *  1 byte         n bytes
 * | header |         body |
 *
 * Where header is a single byte of the form:
 *
 * - header more than 0 indicating that this packet is for the `OUT` stream
 * - header less than 0 indicating that this packet is for the `ERR` stream
 * - abs(header) indicating the length of the packet body, in bytes
 * - header == 0 indicating the end of the stream
 *
 * Writes to either of the two `Output`s are synchronized on the shared
 * `destination` stream, ensuring that they always arrive complete and without
 * interleaving. On the other side, a `Pumper` reads from the combined
 * stream, forwards each packet to its respective destination stream, or terminates
 * when it hits a packet with `header == 0`
 */
public class ProxyStream{

    public static final int OUT = 1;
    public static final int ERR = -1;
    public static final int END = 0;

    public static void sendEnd(OutputStream out) throws IOException {
        synchronized(out){
            out.write(ProxyStream.END);
            out.flush();
        }
    }

    public static class Output extends java.io.OutputStream {
        private java.io.OutputStream destination;
        private int key;

        public Output(java.io.OutputStream out, int key){
            this.destination = out;
            this.key = key;
        }

        @Override public void write(int b) throws IOException {
            synchronized (destination){
                destination.write(key);
                destination.write(b);
            }
        }

        @Override public void write(byte[] b) throws IOException {
            if (b.length > 0) {
                synchronized (destination) {
                    write(b, 0, b.length);
                }
            }
        }

        @Override public void write(byte[] b, int off, int len) throws IOException {

            synchronized (destination) {
                int i = 0;
                while (i < len && i + off < b.length) {
                    int chunkLength = Math.min(len - i, 127);
                    if (chunkLength > 0) {
                        destination.write(chunkLength * key);
                        destination.write(b, off + i, Math.min(b.length - off - i, chunkLength));
                        i += chunkLength;
                    }
                }
            }
        }

        @Override public void flush() throws IOException {
            synchronized (destination) {
                destination.flush();
            }
        }

        @Override public void close() throws IOException {
            synchronized (destination) {
                destination.close();
            }
        }
    }

    public static class Pumper implements Runnable{
        private InputStream src;
        private OutputStream destOut;
        private OutputStream destErr;
        private Object synchronizer;
        public Pumper(InputStream src, OutputStream destOut, OutputStream destErr, Object synchronizer){
            this.src = src;
            this.destOut = destOut;
            this.destErr = destErr;
            this.synchronizer = synchronizer;
        }
        public Pumper(InputStream src, OutputStream destOut, OutputStream destErr){
            this(src, destOut, destErr, new Object());
        }

        public void preRead(InputStream src){}

        public void preWrite(byte[] buffer, int length){}

        public void run() {

            byte[] buffer = new byte[1024];
            while (true) {
                try {
                    this.preRead(src);
                    int header = src.read();
                    // -1 means socket was closed, 0 means a ProxyStream.END was sent. Note
                    // that only header values > 0 represent actual data to read:
                    // - sign((byte)header) represents which stream the data should be sent to
                    // - abs((byte)header) represents the length of the data to read and send
                    if (header == -1 || header == 0) break;
                    else {
                        int stream = (byte) header > 0 ? 1 : -1;
                        int quantity0 = (byte) header;
                        int quantity = Math.abs(quantity0);
                        int offset = 0;
                        int delta = -1;
                        while (offset < quantity) {
                            this.preRead(src);
                            delta = src.read(buffer, offset, quantity - offset);
                            if (delta == -1) {
                                break;
                            } else {
                                offset += delta;
                            }
                        }

                        if (delta != -1) {
                            synchronized (synchronizer) {
                                this.preWrite(buffer, offset);
                                switch(stream){
                                    case ProxyStream.OUT: destOut.write(buffer, 0, offset); break;
                                    case ProxyStream.ERR: destErr.write(buffer, 0, offset); break;
                                }
                            }
                        }
                    }
                } catch (org.newsclub.net.unix.ConnectionResetSocketException e) {
                    // This happens when you run mill shutdown and the server exits gracefully
                    break;
                } catch (IOException e) {
                    // This happens when the upstream pipe was closed
                    break;
                }
            }

            try {
                synchronized (synchronizer) {
                    destOut.flush();
                    destErr.flush();
                }
            } catch(IOException e) {}
        }

        public void flush() throws IOException {
            synchronized (synchronizer) {
                destOut.flush();
                destErr.flush();
            }
        }
    }
}

File: main/client/src/mill/main/client/ProxyStream.scala
package mill.main.client

import java.io.{IOException, InputStream, OutputStream}
import scala.util.control._

object ProxyStream{
  final val OUT: Int = 1
  final val ERR: Int = -1
  final val END: Int = 0


  def sendEnd(out: OutputStream): Unit = {
    // Synchronize on the OutputStream instance to ensure thread safety
    out.synchronized {
      out.write(ProxyStream.END)
      out.flush()
    }
  }


  class Output(destination: OutputStream, key: Int) extends OutputStream {

    override def write(b: Int): Unit = {
      destination.synchronized { // Synchronize on the destination OutputStream
        destination.write(key)
        destination.write(b)
      }
    }

    override def write(b: Array[Byte]): Unit = {
      if (b.nonEmpty) {
        destination.synchronized { // Synchronize on the destination OutputStream
          write(b, 0, b.length)
        }
      }
    }

    override def write(b: Array[Byte], off: Int, len: Int): Unit = {
      destination.synchronized { // Synchronize on the destination OutputStream
        var i = 0
        while (i < len && i + off < b.length) {
          val chunkLength = Math.min(len - i, 127)
          if (chunkLength > 0) {
            destination.write(chunkLength * key)
            destination.write(b, off + i, Math.min(b.length - off - i, chunkLength))
            i += chunkLength
          }
        }
      }
    }

    override def flush(): Unit = {
      destination.synchronized { // Synchronize on the destination OutputStream
        destination.flush()
      }
    }

    override def close(): Unit = {
      destination.synchronized { // Synchronize on the destination OutputStream
        destination.close()
      }
    }
  }

  class Pumper(
      src: InputStream,
      destOut: OutputStream,
      destErr: OutputStream,
      synchronizer: AnyRef = new Object()
  ) extends Runnable {

    def preRead(src: InputStream): Unit = ()

    def preWrite(buffer: Array[Byte], length: Int): Unit = ()

    override def run(): Unit = {
      val buffer = Array.ofDim[Byte](1024)
      var shouldContinue = true

       val outer = new Breaks;

      outer.breakable {
			while (shouldContinue) {
				try {
					preRead(src);
					val header = src.read();
					// -1 means socket was closed, 0 means a ProxyStream.END was sent. Note
					// that only header values > 0 represent actual data to read:
					// - sign((byte)header) represents which stream the data should be sent to
					// - abs((byte)header) represents the length of the data to read and send
					if (header == -1 || header == 0)
						outer.break;
					else {
						val stream = if(header.toByte > 0) 1 else -1;
						val quantity0 = header.toByte;
						val quantity = Math.abs(quantity0);
						var offset = 0;
						var delta = -1;
						while (offset < quantity) {
							preRead(src);
							delta = src.read(buffer, offset, quantity - offset);
							if (delta == -1) {
                shouldContinue = false
								outer.break;
							} else {
								offset += delta;
							}
						}

            // println("buffer: " + buffer.mkString(","))
            // println(delta)
            // println(offset)

						if (delta != -1) {
							synchronizer.synchronized {
								preWrite(buffer, offset);
                // println("stream: " + stream)
								stream match {
                  case ProxyStream.OUT =>
                    // println("out")
                    destOut.write(buffer, 0, offset);
                    // outer.break;
                  case ProxyStream.ERR =>
                    // println("err")
                    destErr.write(buffer, 0, offset);
                    // outer.break;
								}
              }
							}
						}
					}
				 catch {
          case e:IOException =>
            shouldContinue = false
					  outer.break;
				}
			}
    }

    try {
      synchronizer.synchronized {
        destOut.flush()
        destErr.flush()
      }
    } catch {
      case _: IOException => ()
    }
    }

    def flush(): Unit = {
      synchronizer.synchronized {
        destOut.flush()
        destErr.flush()
      }
    }
  }


}
